lib/securenative/event_manager.rb

```diff
@@ -1,158 +1,162 @@
-                     end
-
-      @queue = []
-      @semaphore = Mutex.new
-      @interval = options.interval
-      @options = options
-      @send_enabled = false
-      @attempt = 0
-      @coefficients = [1, 1, 2, 3, 5, 8, 13]
-
-      @thread = Thread.new { run }
+  def initialize(options = Options.new, http_client = nil)
+    if options.api_key.nil?
+      raise SecureNativeSDKError, 'API key cannot be None, please get your API key from SecureNative console.'
     end
 
-    def send_async(event, resource_path)
-      if @options.disable
-        SecureNativeLogger.warning('SDK is disabled. no operation will be performed')
-        return
-      end
+    @http_client = if http_client.nil?
+                     SecureNative::HttpClient.new(options)
+                   else
+                     http_client
+                   end
+
+    @queue = []
+    @semaphore = Mutex.new
+    @interval = options.interval
+    @options = options
+    @send_enabled = false
+    @attempt = 0
+    @coefficients = [1, 1, 2, 3, 5, 8, 13]
+
+    @activated = false
+    @thread = nil
+  end
 
-      item = QueueItem.new(resource_path, EventManager.serialize(event).to_json, false)
-      @queue.append(item)
+  def send_async(event, resource_path)
+    if @options.disable
+      SecureNative::Log.warning('SDK is disabled. no operation will be performed')
+      return
     end
 
-    def flush
-      @queue.each do |item|
-        @http_client.post(item.url, item.body)
+    if @activated == false
+      @activated = true
+      begin
+        @thread = Thread.new { run }
+      rescue StandardError => e
+        SecureNative::Log.error("Could not start event scheduler; #{e}")
       end
     end
 
-    def send_sync(event, resource_path, retry_sending)
-      if @options.disable
-        SecureNativeLogger.warning('SDK is disabled. no operation will be performed')
-        return
-      end
+    item = QueueItem.new(resource_path, EventManager.serialize(event).to_json, false)
+    @queue.append(item)
+  end
 
-      SecureNativeLogger.debug("Attempting to send event #{event}")
-      res = @http_client.post(resource_path, EventManager.serialize(event).to_json)
+  def flush
+    @queue.each do |item|
+      @http_client.post(item.url, item.body)
+    end
+  end
 
-      if res.nil? || res.code != '200'
-        SecureNativeLogger.info("SecureNative failed to call endpoint #{resource_path} with event #{event}. adding back to queue")
-        item = QueueItem.new(resource_path, EventManager.serialize(event).to_json, retry_sending)
-        @queue.append(item)
-      end
+  def send_sync(event, resource_path, retry_sending)
+    if @options.disable
+      SecureNative::Log.warning('SDK is disabled. no operation will be performed')
+      return
+    end
 
-      res
+    SecureNative::Log.debug("Attempting to send event #{event}")
+    res = @http_client.post(resource_path, EventManager.serialize(event).to_json)
+
+    if res.nil? || res.code != '200'
```

Comment:
It's sound like good feature to do retry here, but as i unnderstand verify always called with false, so it will nnever work, and also u should not retry on 4XX errors